### PR TITLE
Add constants and metadata properties for MergeRollupTask:

### PR DESCRIPTION
**Pull request by jtao15**
Part of #2715, add constants and metadata based on [updated design](https://docs.google.com/document/d/1-AKCfXNXdoNjFIvJ87wjWwFM_38gS0NCwFrIYjYsqp8/edit?usp=sharing):
1. Add setter/getter method for merge/rollup watermarks of each granularity.
2. Add task config related constants in MinionConstants, create until functions to get rollup aggregation configs and merge constraints.
3. Add MergeRollupTask.bucketGranularity constant which will be used in customMap of SegmentMetadata.

### CODE REVIEW

**Comment by jtao15** [comment]
@Jackie-Jiang @snleee

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * The <code>watermarkMsMap</code> denotes the time (exclusive) upto which tasks have been executed for the bucket granularity.
+ *
+ * This gets serialized and stored in zookeeper under the path MINION_TASK_METADATA/MergeRollupTask/tableNameWithType
+ */
+public class MergeRollupTaskMetadata {
+
+  private static final String WATERMARK_KEY_PREFIX = "watermarkMs_";
+
+  private final String _tableNameWithType;
+  private final Map<String, Long> _watermarkMsMap;
```
Comment (snleee) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java:39]: `watermarkMap`?

Also, please add the comment on what's the key/value.

File: pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java [hunk 2]
```diff
@@ -58,8 +60,27 @@ private MinionConstants() {
...
+    public static final String TASK_TYPE = "MergeRollupTask";
+
+    public static final String COLLECTOR_TYPE_KEY = "collectorType";
+
+    // Rollup aggregate function related configs
+    public static final String AGGREGATE_KEY_PREFIX = "aggregate";
+
+    // Merge properties related configs
+    public static final String MERGE_KEY_PREFIX = "merge";
+    public enum Granularity {
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java:72]: I think that we should put `Granularity` somewhere else. We can put this somewhere together with `MergeRollupTaskMetadata`

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java (hunk 1 above)
Comment (snleee) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java:39]: Maybe `Map<Granularity, Long>` would be better?

File: pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java [hunk 3]
```diff
@@ -177,7 +177,7 @@ public Builder setSegmentName(String segmentName) {
     }
 
     public MergeRollupSegmentConverter build() {
-      return new MergeRollupSegmentConverter(_inputIndexDirs, _workingDir, _tableName, _segmentName, _mergeType,
+      return new MergeRollupSegmentConverter(_inputIndexDirs, _workingDir, _tableName, _segmentName, _collectorType,
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java:180]: `MergeRollupSegmentConverter` is actually the old code and we should deprecate this. We now need to use `SegmentProcessorFramework`. Can you add the comment? We can address this from another PR. Please refer to `RealtimeToOfflineSegmentsTaskExecutor` on how to use segment processing framework.

**Review by snleee — COMMENTED** [review]
partial review

File: pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java (hunk 2 above)
Comment (jtao15) [reply pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java]: Moved this enum into `MergeRollupTaskMetadata`

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java (hunk 1 above)
Comment (jtao15) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java]: Updated.

File: pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java (hunk 3 above)
Comment (jtao15) [reply pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java:184]: Added the comment for this class.

File: pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java [hunk 4]
```diff
@@ -45,16 +49,16 @@
   private TableConfig _tableConfig;
   private String _tableName;
   private String _segmentName;
-  private MergeType _mergeType;
+  private CollectorType _collectorType;
```
Comment (snleee) [inline pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java:52]: I think that we can just revert these changes back. Your separate PR for migrating from  to `SegmentProcessingFramework` will deprecate the usage of `MergeRollupSegmentConverter`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.pinot.common.minion.MergeRollupTaskMetadata;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.segment.processing.collector.ValueAggregatorFactory;
+import org.apache.pinot.pql.parsers.utils.Pair;
+import org.apache.pinot.spi.utils.TimeUtils;
+
+
+public class MergeRollupTaskUtils {
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:32]: Can we add the unit test for this util class?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        mergePropertiesMap.putIfAbsent(granularity, new HashMap<>());
+        mergePropertiesMap.get(granularity).put(mergeProperty, entry.getValue());
+      }
+    }
+
+    Map<MergeRollupTaskMetadata.Granularity, MergeProperties> allMergeProperties = new HashMap<>();
+    for (Map.Entry<MergeRollupTaskMetadata.Granularity, Map<String, String>> entry : mergePropertiesMap.entrySet()) {
+      Map<String, String> properties = entry.getValue();
+      MergeProperties mergeProperties = new MergeProperties(
+          TimeUtils.convertPeriodToMillis(properties.get(MinionConstants.MergeRollupTask.BUFFER_TIME)),
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:91]: We also need to cover `mergeType` for each gran

File: pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java (hunk 4 above)
Comment (jtao15) [reply pinot-core/src/main/java/org/apache/pinot/core/minion/rollup/MergeRollupSegmentConverter.java:52]: Yes, this will be deprecated, I'll revert these changes back.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java (hunk 6 above)
Comment (jtao15) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:104]: Added

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java (hunk 5 above)
Comment (jtao15) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:33]: Added

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * Metadata for the minion task of type <code>MergeRollupTask</code>.
+ * The <code>watermarkMap</code> denotes the time (exclusive) upto which tasks have been executed for the bucket granularity.
+ *
+ * This gets serialized and stored in zookeeper under the path MINION_TASK_METADATA/MergeRollupTask/tableNameWithType
+ */
+public class MergeRollupTaskMetadata {
+
+  public enum Granularity {
```
[the remark this answers is not in the document]
Comment (jtao15) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MergeRollupTaskMetadata.java]: Put it in a separate file

**Review by snleee — APPROVED** [review]
LGTM

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.segment.processing.collector.CollectorFactory;
+import org.apache.pinot.core.segment.processing.collector.ValueAggregatorFactory;
+import org.apache.pinot.pql.parsers.utils.Pair;
+import org.apache.pinot.spi.utils.TimeUtils;
+
+
+public class MergeRollupTaskUtils {
+
+  public static class MergeProperties {
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:35]: I think that it's better to put this as a separate class. `Util` class should only provide the static helper functions that are commonly used.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    private final CollectorFactory.CollectorType _mergeType;
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java:37]: Also, we should think about having `MergeType` separately.

IMO, `MergeType` concept will be subset of `CollectorType`.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java (hunk 8 above)
Comment (jtao15) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java]: Put it as a separate file.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java (hunk 9 above)
Comment (jtao15) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskUtils.java]: Making a check in `getAllMergeProperties()` util function to make sure only `CONCAT` and `ROLLUP` is allowed.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeProperties.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+  private final CollectorFactory.CollectorType _mergeType;
+  private final long _bufferMs;
+  private final long _maxNumRecordsPerSegment;
+  private final long _maxNumRecordsPerTask;
+
+  public MergeProperties(CollectorFactory.CollectorType mergeType, long bufferMs, long maxNumRecordsPerSegment,
+      long maxNumRecordsPerTask) {
+    _mergeType = mergeType;
+    _bufferMs = bufferMs;
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeProperties.java:34]: `bufferTimeMs`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeProperties.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * under the License.
+ */
+package org.apache.pinot.plugin.minion.tasks.merge_rollup;
+
+import org.apache.pinot.core.segment.processing.collector.CollectorFactory;
+
+
+public class MergeProperties {
+
+  private final CollectorFactory.CollectorType _mergeType;
```
Comment (snleee) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeProperties.java:26]: Let's use `String` for `_mergeType`. We can add a function in util class that allows the conversion from `String mergeType -> CollectorType` and perform validation there.

**Review by snleee — APPROVED** [review]
